lib/mongo/operation/shared/sessions_supported.rb

```diff
@@ -167,6 +167,17 @@ def add_read_preference(sel, connection)
         # https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-single
         if connection.description.standalone?
           # Read preference is never sent to standalones.
+        elsif connection.description.mongos?
+          # When server is a mongos:
+          # - $readPreference is never sent when mode is 'primary'
+          # - When mode is 'secondaryPreferred' $readPreference is only sent
+          #   when a non-mode field (i.e. tag_sets) is present
+          # - Otherwise $readPreference is sent
+          if read && read.to_doc[:mode] == 'secondaryPreferred'
+            sel['$readPreference'] = read.to_doc unless read.to_mongos.nil?
+          elsif read && read.to_doc[:mode] != 'primary'
+            sel['$readPreference'] = read.to_doc
+          end
```

Comment:
The " # In replica sets and sharded clusters, read preference is passed" comment below this hunk should be updated I would think.